wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/resource/AbstractFileUploadResource.java

```diff
@@ -0,0 +1,223 @@
+import org.apache.wicket.protocol.http.servlet.ServletWebRequest;
+import org.apache.wicket.request.cycle.RequestCycle;
+import org.apache.wicket.request.resource.AbstractResource;
+import org.apache.wicket.util.lang.Bytes;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import com.github.openjson.JSONObject;
+
+/**
+ * The resource that handles the file uploads.
+ * Reads the file items from the request parameters and uses {@link IUploadsFileManager}
+ * to store them.
+ * Additionally, cares about the response's content type and body.
+ * <p>
+ * This code was adapted from
+ * <p>
+ * <a href="https://github.com/martin-g/blogs/blob/master/file-upload/src/main/java/com/mycompany/fileupload/AbstractFileUploadResource.java">AbstractFileUploadResource.java</a>
+ * <p>
+ * The main difference is that there some JQuery plugin is used at client side (and it supports multiple uploads +
+ * some UI allowing to delete/preview files and so on).
+ * Here we are just using plain jQuery code at client side to upload a single file.
+ */
+public abstract class AbstractFileUploadResource extends AbstractResource
+{
+	private static final Logger LOG = LoggerFactory.getLogger(AbstractFileUploadResource.class);
+
+	public static final String PARAM_NAME = "WICKET-FILE-UPLOAD";
+
+	/**
+	 * This resource is usually an application singleton. Thus, client side pass
+	 * to the resource a unique ID identifying the upload field performing the upload.
+	 * The upload file makes sure this is a unique identifier at application level, See
+	 * {@link FileUploadToResourceField#generateAUniqueApplicationWiseId()}. So that, there are no clashes between
+	 * different users/pages/sessions performing an upload.
+	 */
+	public static final String UPLOAD_ID = "uploadId";
+	/**
+	 * i18n key for case no files were selected.
+	 */
+	public static final String NO_FILE_SELECTED = "wicket.no.files.selected";
+	/**
+	 * i18n key for the case selcted files exceed size limit.
+	 */
+	public static final String REQUEST_SIZE_LIMIT_EXCEEDED = "wicket.multipart.size.exceeded";
+
+	private final IUploadsFileManager fileManager;
+
+	public AbstractFileUploadResource(IUploadsFileManager fileManager)
+	{
+		this.fileManager = fileManager;
+	}
+
+	/**
+	 * Reads and stores the uploaded files
+	 *
+	 * @param attributes
+	 *            Attributes
+	 * @return ResourceResponse
+	 */
+	@Override
+	protected ResourceResponse newResourceResponse(Attributes attributes)
+	{
+		final ResourceResponse resourceResponse = new ResourceResponse();
+
+		final ServletWebRequest webRequest = (ServletWebRequest) attributes.getRequest();
+
+		// get the ID of the upload field (it should be unique per application)
+		String uploadId = webRequest.getRequestParameters().getParameterValue(UPLOAD_ID).toString("resource");
+
+		try
+		{
+			MultipartServletWebRequest multiPartRequest = webRequest.newMultipartWebRequest(getMaxSize(), uploadId);
```

Comment:
FileCountLimitExceededException exception should also be handled,

Please see ade93326346e20587242b501422e43ae6937a843

Reply:
Ok. Then probably the resource needs a setting for maxFiles too